src/promptflow/promptflow/_core/openai_injector.py

```diff
@@ -81,50 +121,103 @@ def get_aoai_telemetry_headers() -> dict:
 
 
 def inject_operation_headers(f):
-    @functools.wraps(f)
-    def wrapper(*args, **kwargs):
+    def inject_headers(kwargs):
         # Inject headers from operation context, overwrite injected header with headers from kwargs.
         injected_headers = get_aoai_telemetry_headers()
-        original_headers = kwargs.get("headers") if IS_LEGACY_OPENAI else kwargs.get("extra_headers")
+        original_headers = kwargs.get("headers" if IS_LEGACY_OPENAI else "extra_headers")
         if original_headers and isinstance(original_headers, dict):
             injected_headers.update(original_headers)
-        kwargs.update(headers=injected_headers) if IS_LEGACY_OPENAI else kwargs.update(extra_headers=injected_headers)
+        kwargs["headers" if IS_LEGACY_OPENAI else "extra_headers"] = injected_headers
 
-        return f(*args, **kwargs)
+    if asyncio.iscoroutinefunction(f):
+
+        @functools.wraps(f)
+        async def wrapper(*args, **kwargs):
+            inject_headers(kwargs)
+            return await f(*args, **kwargs)
+
+    else:
+
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            inject_headers(kwargs)
+            return f(*args, **kwargs)
 
     return wrapper
 
 
-def inject(f):
-    wrapper_fun = inject_operation_headers((inject_function(["api_key", "headers"])(f)))
+def inject_async(f):
+    wrapper_fun = inject_operation_headers((inject_function_async(["api_key", "headers", "extra_headers"])(f)))
+    wrapper_fun._original = f
+    return wrapper_fun
+
+
+def inject_sync(f):
+    wrapper_fun = inject_operation_headers((inject_function_sync(["api_key", "headers", "extra_headers"])(f)))
     wrapper_fun._original = f
     return wrapper_fun
 
 
-def available_openai_apis():
+def _openai_api_list(sync=True):
```

Comment:
Maybe we can directly return a tuple list? (sync_apis, async_apis)

Reply:
Seems that user can not tell which is sync and which is async from the returning.

Reply:
Return a tuple with injector together.